### ISSUE TITLE
perf: stop copying every per-expert tensor before stacking it

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#948]
## Problem / Background

`stack_individual_experts_with_count` (`src/models/switch_layers.rs:388`) joins a checkpoint's per-expert projection tensors into the stacked `[num_experts, ...]` layout the gather paths expect. It calls `mlxcel_core::copy` on every single tensor:

```rust
while let Some(weight) = weights.get(&expert_key(idx, "weight")) {
    stacked_weight.push(mlxcel_core::copy(weight));
    if has_scales { stacked_scales.push(mlxcel_core::copy(weights.get(&expert_key(idx, "scales"))?)); }
    if has_biases { stacked_biases.push(mlxcel_core::copy(weights.get(&expert_key(idx, "biases"))?)); }
    idx += 1;
}

let weight = mlxcel_core::stack_owned(&stacked_weight, 0);
```

The copies exist only to produce owned `UniquePtr<MlxArray>` values for `stack_owned`, which takes `&[UniquePtr<MlxArray>]`. They contribute nothing to the result: `mlxcel_core::copy` lowers to `mx::copy` (`mlx/ops.cpp:303`), a `Copy` primitive that allocates and writes a full-size output, and `mx::stack` (`mlx/ops.cpp:1336`) then expands and concatenates into an output buffer it allocates itself. Every byte written by the `Copy` is read once by the `Concatenate` and discarded.

### The cost is at first forward, not at load

Both ops are lazy. `mx::copy` and `mx::stack` build graph nodes and allocate nothing, and nothing evaluates model parameters between construction and the first forward pass: the `eval_all` calls in `src/models/sanitize.rs` run on the `WeightMap` before `from_weights` is reached, and the model constructor adds no `eval`. So the wasted work is not load-time allocation; it is copy traffic materialized when the first forward pass evaluates the weight graph, plus a transient allocation of one projection's worth of expert copies while the concatenate runs.

### Volume for Ling-lite-1.5

Substantiated from `models/ling-lite-1.5/model.safetensors.index.json` plus the four shard headers, without reading tensor data. Re-derived on `main` at `a3c823611`; every figure below is exact:

- 5603 tensors total, of which 5376 are routed per-expert tensors (`model.layers.{0..27}.mlp.experts.{0..63}.{gate,up,down}_proj.weight`), matching 64 experts times 3 projections times 28 layers. The checkpoint carries no `switch_mlp` key, so every one of them goes through this branch.
- All BF16. 3584 of them are `[1408, 2048]` and 1792 are `[2048, 1408]`; each is 5,767,168 bytes.
- Total per-expert payload: **31,004,295,168 bytes (31.0 GB, 28.9 GiB)** out of the checkpoint's 33,603,948,544 byte `total_size`.

So the first forward pass writes and re-reads roughly 31 GB that it did not need to touch. The doc comment at `src/models/bailing_moe.rs:1123-1128` already records the 5376 figure for this checkpoint.

### Scope

This is shared code, not a Bailing-specific path. `SwitchLinear::from_weights_with_mode` (`src/models/switch_layers.rs:236`) is the single loader behind the shared `SwitchGLU`.

Nineteen model files reach that loader, either through the shared `SwitchGLU` or by using the shared `SwitchLinear` directly: `bailing_moe.rs`, `cohere2_moe.rs`, `dots1.rs`, `gemma4.rs`, `granitemoehybrid.rs`, `kimi_linear.rs`, `lfm2.rs`, `llada2_moe/mod.rs`, `longcat_flash_ngram.rs`, `mellum.rs`, `minimax.rs`, `minimax_m3_moe.rs`, `mistral4.rs`, `mixtral.rs`, `moondream3.rs`, `olmoe.rs`, `phimoe.rs`, `qwen2_moe.rs`, and `solar_open.rs`. Any of them hits this branch whenever its checkpoint ships experts unstacked. `deepseek.rs:857` also calls `stack_individual_experts` directly for the `baidu/Unlimited-OCR` per-expert checkpoint, which makes twenty entry points in total.

A correction to the original framing of this section: DeepSeek V3.2 is not the lone exception. Fifteen or so model files define their own local `SwitchLinear` / `SwitchGLU` rather than using the shared one, and none of them reaches this branch: `deepseek.rs`, `deepseek_v2.rs`, `deepseek_v3.rs`, `deepseek_v32.rs`, `ernie4_5_moe.rs`, `exaone_moe.rs`, `glm4_moe.rs`, `glm4_moe_lite.rs`, `hunyuan_moe.rs`, `jamba.rs`, `llama4.rs`, `qwen3_moe.rs` (and `qwen3_vl_moe.rs`, which imports its types), `qwen3_next.rs` (and `qwen3_5.rs`, which reuses them), plus `step3p5.rs` and `gpt_oss.rs` with differently named local types (`Step3p5SwitchGLU`, `GptOssSwitchGLU`). So Qwen3-MoE, Qwen3-VL-MoE, Qwen3-Next, Qwen3.5, DeepSeek v1/v2/V3/V3.2, GLM4-MoE, GLM4-MoE-Lite, ERNIE 4.5 MoE, ExaOne MoE, Hunyuan-MoE, Llama 4, Jamba, gpt-oss and Step3p5 are all out of scope for the per-expert branch, not just DeepSeek V3.2. Note that none of those local implementations has a per-expert stacking branch at all: they only handle the already-stacked layout, so an unstacked checkpoint for those families would fail to load rather than take a slow path.

Ling-lite is the largest known instance because its published checkpoint is fully unstacked, not because Bailing is special.

The same redundant `Copy` sits three lines above, on the pre-stacked branch (`switch_layers.rs:245-253`), where `copy` is applied to the already-joined `[num_experts, ...]` weight, scales, and biases. That is the common case for mlx-community conversions and it is the same byte volume, so it is worth fixing in the same pass, but it needs a different mechanism (see below). That second instance is far more widespread than the first: every one of the local `SwitchLinear` implementations listed above copies the joined tensor the same way (`qwen3_moe.rs:983-987` and its `get_weight_copy` helper at `:1005` are representative), so a handle-sharing shim would pay off well beyond `switch_layers.rs`.

Both branches predate #946 / #838; the Bailing MoE work is only where this was noticed.

## Proposed Solution

For the per-expert branch, no new entry point is needed. `ops::stack_owned` (`src/lib/mlxcel-core/src/ops.rs:44`) does nothing but map the owned buffers to `*const ffi::MlxArray` and call `ffi::stack(&ptrs, axis)`, and the pointer-taking `ops::stack` (`ops.rs:37`, re-exported at `lib.rs:2592`) is already public. The C++ side (`stack` in `src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp:2577`) copies each `a->inner` into its own `std::vector<array>`, and `mx::array` is a refcounted handle, so the stacked graph node keeps its inputs alive independently of the `WeightMap`. Collecting `*const MlxArray` straight from `weights.get(...)` and calling `mlxcel_core::stack` is sufficient and safe here.

One in-tree precedent does exactly this job: `src/models/olmoe.rs:566-582`, which drains the source entries with `weights.remove(&key)`, collects `*const MlxArray` from the owned values, and calls `mlxcel_core::stack` with no `copy` at all.

`src/models/phimoe.rs` is NOT a second precedent, contrary to the original text of this issue. It builds the same pointer vector and calls `mlxcel_core::stack` at `:526`, but it populates `expert_arrays` with `mlxcel_core::copy(w)` at `:517` first, so it carries the identical redundant `Copy` this issue is about, over its own expert set. It belongs in the scope of the fix rather than in the list of things to copy from.

The pre-stacked branch is the one that would need something new. Turning `copy(weight)` into a borrow requires a bridge shim that wraps the same refcounted handle (`std::make_unique<MlxArray>(a.inner)`, no primitive), since `mlxcel_core::copy` is currently the only way to get an owned `UniquePtr` from a `&MlxArray`. That shim does not exist yet. It is a few lines and it would also cut the same waste out of every other `get_weight_copy`-style loader, so it is worth scoping deliberately rather than folding in silently.

## Acceptance Criteria

- [x] `stack_individual_experts_with_count` no longer calls `mlxcel_core::copy` on per-expert tensors; the stack is built from borrowed pointers.
- [x] The stacked result still owns its inputs independently of the `WeightMap`, verified by a test that drops the source map before evaluating the stacked array.
- [x] The expert-count return value and the truncated-checkpoint error in `stack_individual_experts` are unchanged, with the existing contiguous-gather-until-first-gap semantics preserved.
- [ ] Numerics are identical, verified by a token-exact comparison on at least one real MoE checkpoint that takes the per-expert branch (`models/ling-lite-1.5` is the reference case), not by shape assertions alone. Same prompt, same seed, same token ids before and after. **Not run in PR #960**: the release binaries are not built in that environment and the reference checkpoint is 33.6 GB. Left for the follow-up validation pass.
- [ ] A second real checkpoint from a different family that also takes this branch (a per-expert Mixtral or Qwen2-MoE checkpoint) is smoke-tested, since the loader is shared. **Not run in PR #960**, same reason.
- [ ] Measured and reported: first-forward wall time and peak memory on `models/ling-lite-1.5` before and after, so the saved traffic is quantified rather than assumed. **Not run in PR #960**, same reason. Peak should differ measurably even though steady-state resident size does not, because the per-projection transient (about 352 MiB for Ling-lite) goes away with the copies.
- [x] The same redundant copy in `src/models/phimoe.rs:517` is either fixed alongside it or split into a follow-up with the reason recorded.
- [x] A decision is recorded on the pre-stacked branch at `switch_layers.rs:245-253`: either fixed in the same change with a handle-sharing bridge shim, or split into a follow-up issue with the reason. If a shim is added, note that the same pattern exists in every local `SwitchLinear` implementation listed under Scope, so the follow-up should say whether those are in or out.

## Technical Considerations

- Lifetime: the pointers must be built while the `&WeightMap` borrow is live and the map is not mutated in between. The current function only reads, so this holds; keep it that way rather than switching to `remove`-based draining as `olmoe.rs` does, because `stack_individual_experts` is called from a `&WeightMap` context.
- Peak memory today is bounded by one projection's expert set held live while the concatenate runs (64 times 5.5 MiB, about 352 MiB, for Ling-lite), not by the full 31 GB, because MLX frees each `Copy` output once the `Concatenate` consumes it. Removing the copies removes that transient too, so the before and after peak numbers should differ measurably even though the steady-state resident size does not change.
- `mlxcel_core::stack` takes raw pointers and is declared safe while performing an `unsafe` FFI call, so keep the pointer collection adjacent to its use and do not let it outlive the borrow.

<details>
<summary><b>Implementation context (machine facts, validation method, and a recurring defect class)</b></summary>

**Most relevant to this issue:** the machine build facts and the token-exact reference oracle. The acceptance criteria above require proving identical numerics on a real MoE checkpoint rather than by shape tests, and the oracle section below is the method for that; the build notes cover the `DEVELOPER_DIR` requirement and the narrow-selector rule that any local verification run here depends on.

Collected while porting five text-model families in #924, #926, #928, #930, and #946. This
is background for whoever picks this up, not additional scope.

### Build and test on the current development machine (Apple M1 Ultra, macOS)

- Every cargo invocation needs `DEVELOPER_DIR=/Applications/Xcode-26.6.0.app/Contents/Developer`, or cmake fails with `xcrun: error: unable to find utility "metal"`. `xcode-select` still points at CommandLineTools.
- Only the RELEASE profile is warm. A debug `cargo check` triggers a COLD MLX C++ build that takes many minutes. Use `DEVELOPER_DIR=... cargo check --release --lib --tests --features metal,accelerate`.
- Narrow test selectors only: `DEVELOPER_DIR=... cargo test --release --lib --features metal,accelerate models::<module>`. A bare `cargo test`, `cargo test --lib`, or `cargo clippy --workspace --all-targets` runs long enough to stall an agent with a stream-idle watchdog, and a cold `cargo build --release` can exceed it on the MLX C++ link step alone.
- `cargo fmt --all -- --check`, never a bare-path fmt check, which produces spurious edition-2024 diffs.
- CI runs no `cargo test` at all (only `changes`, `cargo-deny`, `cargo-fmt`, and the cross-repo-ref guard). Local runs plus a real checkpoint are the only gate. Two deterministically failing tests were found sitting on `main` during this series for exactly that reason: see #939, and the `FAMILY_ORDER` entries fixed in #946.
- `tracing::warn!` is a NO-OP in the `mlxcel` CLI binary; only `src/server/startup.rs` installs a subscriber. Use `eprintln!` for any CLI-facing diagnostic.
- A test that trips an MLX C++ throw aborts the whole test binary with SIGABRT rather than failing cleanly. That is expected, and the abort is stronger evidence than a clean assertion failure. One SIGABRT on the first run after a fresh link that does not reproduce on reruns is a cold Metal-initialization race, observed independently by two reviewers.

### Validating model behavior: a token-exact reference oracle

Shape tests do not catch the failures that matter here. A wrong prefill/decode offset, a flat instead of interleaved QKV split, the wrong RoPE convention, or a routing bias applied to the wrong copy all produce correctly-shaped tensors and fluent, plausible output. Only the token id sequence separates them.

What worked: create a scratch venv, `pip install mlx-lm`, and drive `mlx_lm.generate.stream_generate` with `make_sampler(temp=0.0)`, printing both text and token ids. Then compare against `mlxcel generate`. **Pass `--no-chat-template` to mlxcel even when the checkpoint ships a chat template**, so both sides see the same raw prompt, and run the templated path separately as a usability check.

Caveat: the oracle is blind wherever mlxcel and mlx-lm make the same choice. GELU was one such case, closed by monkey-patching the reference to the other variant and regenerating. Note `mlxcel_core::utils::gelu_approx` is erf-based despite its name, while MLX Python's `nn.gelu_approx` really is the tanh form.

### A recurring defect class worth checking for

A value from `config.json` passes every Rust-side check, violates an undocumented precondition of an MLX C++ entry point, and kills the process at the FIRST FORWARD PASS rather than failing at load. Most cxx bridge functions are declared returning `UniquePtr<MlxArray>` rather than `Result`, so the C++ throw is an uncatchable `std::terminate`, not something `catch_unwind` contains. The model loads cleanly and the server dies on its first request. This class produced one CRITICAL and five HIGH findings across the five ports.

Specifics established by reading the pinned MLX checkout rather than assuming:

- Gathers do NOT range-check positive indices. `take` wraps negative indices but an out-of-range positive index silently returns values belonging to no row, and that reaches the logits with no fault. Bound every gather by the real tensor shape, never by a config field.
- `slice` CLAMPS an out-of-range stop instead of throwing, so a too-wide split silently loses trailing channels.
- `fast_rope` requires `dims` even, positive, and no larger than the last axis.
- `rms_norm` and `layer_norm` never inspect `eps`. A NaN or negative eps yields NaN hidden states with no error at all, which is harder to diagnose than a crash.
- `matmul` throws on an inner-dimension mismatch, which is why an unchecked INPUT axis is fatal even where the row axis bounds only an argmax.
- `quantized_matmul` divides by `bits`, so `"bits": 0` is a division by zero. On AArch64 the divide returns 0 and `std::invalid_argument` fires; on x86-64 the hardware raises `SIGFPE` and kills the process before any exception exists.
- Reconstruct a quantized input width the way MLX does, as `scales.shape(-1) * group_size`, and check `.biases` shape equals `.scales` shape. A self-consistency check cannot catch a checkpoint honestly packed for the wrong `hidden_size`, because such a checkpoint IS self-consistent.
- Put zero checks BEFORE divisibility checks, because `0.is_multiple_of(0)` is true.
- Avoid unbounded probe loops over `0..n_layer` in the load path: a huge `n_layer` hangs with a flat allocation footprint, so no OOM kill rescues it.

`src/models/gpt_neox.rs` and `src/models/helium.rs` carry the current `ModelArgs::validate` / `validate_weights` shape to copy from.

### Conventions that bite

- The `// Used by:` comments above shared helpers are the designated discovery mechanism for "what breaks if I change this", and they can themselves be stale. One was found listing six callers where there were eight. Verify by grep rather than trusting the comment, and update it when you touch the helper.
- `mlxcel arch` is the architecture registry. `mlxcel list` lists downloaded models and will not tell you whether a family is supported.

</details>

**Comment by inureyes** [comment]
Correction to this issue's cost model, established while implementing it in #960 and confirmed against the pinned MLX checkout.

The redundant `Copy` is real and #960 removes it, but it is not the cost this issue describes. `mlxcel_core::copy` does not materialize. `mx::copy` builds a `Copy` node (`mlx/ops.cpp:303`); `Copy::eval_gpu` (`mlx/backend/gpu/primitives.cpp:64`) and `Copy::eval_cpu` (`mlx/backend/cpu/primitives.cpp:88`) both delegate to `Copy::eval`; and `Copy::eval` (`mlx/backend/common/common.cpp:55`) is `out.copy_shared_buffer(inputs[0])`, which aliases the input buffer. It allocates nothing and writes nothing. The materializing `copy_gpu(..., CopyType::General)` sitting a few lines above it belongs to `Contiguous::eval_gpu`, a different primitive, and that adjacency is the likely source of the original reading.

Consequences for the text above:

- The 31,004,295,168 bytes of first-forward copy traffic for `models/ling-lite-1.5` was never being spent. The tensor inventory itself (5376 per-expert BF16 tensors, 3584 of `[1408, 2048]` and 1792 of `[2048, 1408]`) is correct; only the conclusion drawn from it is not.
- The "Technical Considerations" note about a per-projection transient of about 352 MiB held live while the concatenate runs does not hold either. The `Copy` outputs never had their own buffers, so there is no transient to remove and peak memory should be unchanged rather than improved.
- What #960 actually saves is one graph node per per-expert tensor: graph metadata and eval-scheduling overhead, not memory traffic.

The change is still correct and still worth having, so #960 lands as the fix. The acceptance criterion that would have caught this before it reached a source comment is the before/after measurement one, which is the reason it is worth keeping in the list even when a change looks obviously free.

The pre-stacked follow-up #959 was filed on the same premise and has been closed for the same reason.
